### PR TITLE
Optimize DetectBest execution

### PR DESCRIPTION
**Pull request by moredure**
@unknwon, please review. Should perform better with less memory allocations, although you can check out my fork https://github.com/moredure/chardet for non concurrent version.

### CODE REVIEW

**Comment by moredure** [comment]
Should perform better and less memory allocations

**Comment by unknwon** [comment]
Thanks for the PR!

Can you give more background about why this is going to faster and provide some benchmarks?

**Comment by moredure** [comment]
hi @unknwon , the only change here is that we don't allocate slice for holding and sorting detected values in order to get the optimal, we just reading values until find optimal

**Comment by unknwon** [comment]
This PR also adds code to spawn goroutines which I'm not convinced this is a good idea within this function :)

**Comment by moredure** [comment]
current DetectBest implementation is based on DetectAll which is spawning goroutines

**Comment by moredure** [comment]
If you are interested in removing goroutine creation for this and other APIs i think it's better to do it in separate pull requests with breaking changes to internals for matchHelper to getting rid of channels and goroutines, such as this https://github.com/moredure/chardet

**Comment by unknwon** [comment]
Gotcha, thanks for the context!

If we want to make progress for this PR, I need to understand:

1. how less memory allocation claimed make actual performance improvements, and what's the proof of this conclusion?
2. What's stopping us from changing `DetectAll` directly?

**Comment by moredure** [comment]
DetectAll returns all possible encodings fits data slice sorted by confidence, while DetectBest returns single one with best/higher possible confidence across all detected. in case of DetectBest we don't need any sorting, and that's why we don't need to allocate slice to hold all detected encodings, it's better than just DetectAll and return first. I will share benchmarks soon.

**Comment by moredure** [comment]
@unknwon 
```
goos: darwin
goarch: amd64
pkg: github.com/gogs/chardet
cpu: Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
BenchmarkDetectBest
BenchmarkDetectBest/old
BenchmarkDetectBest/old-4         	    4834	    335575 ns/op	   15444 B/op	      48 allocs/op
BenchmarkDetectBest/new
BenchmarkDetectBest/new-4         	    5865	    199170 ns/op	   13428 B/op	      46 allocs/op
PASS

Process finished with the exit code 0
```

**Comment by moredure** [comment]
Non concurrent detectBest make only 5 allocation but has worser ns/op

**Comment by moredure** [comment]
removing byteStats and converting it into byte array also cuts one more allocation, for both detectBest and detectAll

File: recognizer.go [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (unknwon) [inline recognizer.go]:
```suggestion
```

File: detector_test.go [hunk 2]
```diff
@@ -58,3 +59,28 @@ func TestDetector(t *testing.T) {
...
+			detectBest(textDetector, aaaa)
+		}
+	})
+	b.Run("new", func(b *testing.B) {
+		b.ReportAllocs()
+		for i := 0; i < b.N; i++ {
+			textDetector.DetectBest(aaaa)
+		}
+	})
+}
```
Comment (unknwon) [inline detector_test.go]: This file does not look like had been ran over by "gofmt", could you do that?

File: detector_test.go [hunk 3]
```diff
@@ -58,3 +59,28 @@ func TestDetector(t *testing.T) {
...
+func BenchmarkDetectBest(b *testing.B) {
+	textDetector := chardet.NewTextDetector()
+	aaaa := bytes.Repeat([]byte("A"), 1024)
+	b.Run("old", func(b *testing.B) {
+		b.ReportAllocs()
+		for i := 0; i < b.N; i++ {
+			detectBest(textDetector, aaaa)
+		}
+	})
+	b.Run("new", func(b *testing.B) {
```
Comment (unknwon) [inline detector_test.go]: Thanks for the benchmark comparison! To commit the PR, please clean up/remove old implementation here and only leave the benchmark code for the latest implementation.

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the follow up, I think we can merge this PR after some clean up.

**Review by unknwon — APPROVED** [review]
Thanks for the follow up!

**Pull request merged**